### PR TITLE
[WIP] Quest Events

### PR DESCRIPTION
**Pull request by evilmoses**
## Description
Ability to process quests sent to webhook.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (would cause existing functionality to change)

## Motivation and Context
Since scans are down nobody has worked on including quests in PA. My map now sends a quest webhook when a user submitted a new quest found so needed to process these in PA. Because these quests are not sent by scans, the information that we can send to the webhook is quite limited. Basically, the stop information and the text of the quest description and reward is sent. The expiry is dynamically calculated (23:59 tonight). Because it is just text that is sent, those requiring localization will need to ensure that the information is sent ALREADY TRANSLATED.

## How Has This Been Tested?
Has been tested and has been running on my prod environment for weeks, BUT......

My fork has not been updated to latest patch and is therefore behind the dev branch here. I have tested using latest for about 10 quests without any filters and it works, but needs serious further testing. Also, the following needs doing:

1. caching
2. filter testing (not tested at all)
3. unit tests
4. documentation

### CODE REVIEW

File: PokeAlarm/Filters/QuestFilter.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        self.max_dist = self.evaluate_attribute(  # f.max_dist <= m.distance
+            event_attribute='distance', eval_func=operator.ge,
+            limit=BaseFilter.parse_as_type(float, 'max_dist', data))
+            
+        # Quest Description
+        self.quest_contains = self.evaluate_attribute(  # f.gn matches e.gn
+            event_attribute='quest', eval_func=GymUtils.match_regex_dict,
+            limit=BaseFilter.parse_as_set(
+                GymUtils.create_regex, 'quest_contains', data))
+        self.gym_name_excludes = self.evaluate_attribute(  # f.gn no-match e.gn
```
Comment (LegitDongo) [inline PokeAlarm/Filters/QuestFilter.py]: I thought quests only came from pokestops?
Why does this reference gyms instead of stops?
Glad you're adding filtering by stop name though 👍

**Review by LegitDongo — COMMENTED** [review]
I haven't done any testing yet, just throwing out what I've seen off-hand.

File: PokeAlarm/Filters/QuestFilter.py (hunk 1 above)
Comment (evilmoses) [reply PokeAlarm/Filters/QuestFilter.py]: quests do only come from stops. It only references to contains function in the gym_utils as didn't find it anywhere else, lazy I know.

I haven't added filtering by stop name?? Only quest description and reward description.

The filtering needs more work, I don't use it (not enough user submissions to bother), and just added a couple of filters last minute for those that might need it - no idea if they even work....

I did this in about 30 minutes to get quest reporting working on my server. Needs more work I know, but people kept asking me for it, so easier to PR and let people fill in the blanks.

Comment (LegitDongo) [reply PokeAlarm/Filters/QuestFilter.py]: I should have read a little more, and I need to start drinking coffee, my bad -- `'quest_contains'` 😅 
I wasn't expecting it to work the same way as it would by stops, but it makes sense now that I think about it
We will add more filters later once we actually have it merged, so let's focus on getting this ship-shape.
Something we should consider (maybe in the future if @kvangent doesn't deem it necessary now) is moving create_regex (as well as the other regex functions for good measure) into the GenUtils file since it will be used for more than just gyms, and to eliminate my confusion
Likewise, we will probably want to add some specific names for quests, since using `gym_name_excludes` doesn't really fit the bill - maybe make something more generic? such as a `contains` and `excludes` function somewhere

Comment (evilmoses) [reply PokeAlarm/Filters/QuestFilter.py]: Oops, gym_name_excludes shouldn't be there lol

Comment (evilmoses) [reply PokeAlarm/Filters/QuestFilter.py]: Gone.

**Comment by nileplumb** [comment]
Can't wait this will hopefully encourage the few map forks that support report quest into some sort of order.
Could you also please document everything to the readthedocs / wiki too? Thanks!

**Comment by kvangent** [comment]
Does the quest reward change? If it's sent with each webhook why are we trying to cache it?

**Comment by evilmoses** [comment]
@kvangent, reason for caching is to prevent duplication. Because it is user submitted data, I am finding multiple users submitting the same quest by accident, so want to prevent duplicate posts making it to discord. Just need to check that the same quest/reward has not been sent on same day already etc.

File: PokeAlarm/Manager.py [hunk 2]
```diff
@@ -946,6 +979,56 @@ def process_weather(self, weather):
...
+
+        # Calculate distance and direction
+        if self.__location is not None:
+            quest.distance = get_earth_dist(
+                [quest.lat, quest.lng], self.__location, self.__units)
+            quest.direction = get_cardinal_dir(
+                [quest.lat, quest.lng], self.__location)
+
+        # Store copy of cache info
+        # TODO
```
Comment (kvangent) [inline PokeAlarm/Manager.py]: I think you'll want something like the following: 

```
previous_reward = self._cache.quest_reward(quest.stop_id)
if quest.reward == previous_reward:
    self._log.debug("Quest ignored: Reward previously alerted!")
    return

self._cache.quest_reward(stop_id, quest.reward)
```

But it's worth noting it the Reward is a string, then it'll probably be wrong a lot. If possible we should try to represent the Reward as an enum.

File: PokeAlarm/Events/QuestEvent.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+class QuestEvent(BaseEvent):
+    """ Event representing the discovery of a PokeStop. """
+
+    def __init__(self, data):
+        """ Creates a new Quest Event based on the given dict. """
+        super(QuestEvent, self).__init__('quest')
+
+        # Identification
+        self.stop_id = data['pokestop_id']
+        self.stop_name = data['name']
```
Comment (LegitDongo) [inline PokeAlarm/Events/QuestEvent.py]: It seems a bit ambiguous to have `name` come from the webhook when `name` can also be interpreted as `quest`.
I suggest renaming to `pokestop_name` to match `pokestop_id`.

**Review by LegitDongo — COMMENTED** [review]
Caught this

File: PokeAlarm/Alarms/Discord/DiscordAlarm.py [hunk 4]
```diff
@@ -80,6 +80,15 @@ class DiscordAlarm(Alarm):
             "title": "The weather has changed!",
             "url": "<gmaps>",
             "body": "The weather around <lat>,<lng> has changed to <weather>!"
+        },
+        'quest': {
```
Comment (LegitDongo) [inline PokeAlarm/Alarms/Discord/DiscordAlarm.py]: We should stick with the plural version -- `quests` -- since everything else uses the plural version

File: PokeAlarm/Alarms/Discord/DiscordAlarm.py [hunk 5]
```diff
@@ -113,6 +122,8 @@ def __init__(self, mgr, settings, max_attempts, static_map_key):
             settings.pop('raids', {}), self._defaults['raids'])
         self.__weather = self.create_alert_settings(
             settings.pop('weather', {}), self._defaults['weather'])
+        self.__quest = self.create_alert_settings(
+            settings.pop('quest', {}), self._defaults['quest'])
```
Comment (LegitDongo) [inline PokeAlarm/Alarms/Discord/DiscordAlarm.py]: Same as above

**Review by LegitDongo — COMMENTED** [review]
Specified for DiscordAlarm but should apply for the other types

File: PokeAlarm/Alarms/Pushbullet/PushBulletAlarm.py [hunk 6]
```diff
@@ -50,6 +50,12 @@ class PushbulletAlarm(Alarm):
             'title': "The weather has changed!",
             'url': "<gmaps>",
             'body': "The weather around <lat>,<lng> has changed to <weather>!"
+        },
+        'quests': {
+            'message': "*New quest for <reward>*\n"
```
Comment (kvangent) [inline PokeAlarm/Alarms/Pushbullet/PushBulletAlarm.py]: s/message/title

title -> `"*New quest for  <reward>*"`
body -> `"<quest>"`

File: PokeAlarm/Alarms/FacebookPage/FacebookPageAlarm.py [hunk 7]
```diff
@@ -77,6 +77,15 @@ class FacebookPageAlarm(Alarm):
             'description': "The weather around <lat>,<lng>"
                            " has changed to <weather>!",
             'caption': None
+        },
+        'quests': {
+            'message': "*New quest for <reward>*\n"
```
Comment (kvangent) [inline PokeAlarm/Alarms/FacebookPage/FacebookPageAlarm.py]: To fit in with the other events in this alarm, split message/description on the new line and put the first half in message and second half in description.

File: PokeAlarm/Alarms/Discord/DiscordAlarm.py [hunk 8]
```diff
@@ -80,6 +80,15 @@ class DiscordAlarm(Alarm):
...
             "body": "The weather around <lat>,<lng> has changed to <weather>!"
+        },
+        'quests': {
+            'username': "Quest",
+            'content': "",
+            'icon_url': get_image_url("regular/quest/<type_id>.png"),
+            'avatar_url': get_image_url("regular/quest/<type_id>.png"),
+            'title': "New Quest Found!",
+            'url': "<gmaps>",
+            'body': "Quest will expire at midnight."
```
Comment (kvangent) [inline PokeAlarm/Alarms/Discord/DiscordAlarm.py]: Is there a time remaining DTS we could use here?

File: PokeAlarm/Events/QuestEvent.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+# 3rd Party Imports
+# Local Imports
+from PokeAlarm import Unknown
+from . import BaseEvent
+from PokeAlarm.Utils import get_gmaps_link, get_applemaps_link, \
+    get_waze_link, get_dist_as_str
+
+
+class QuestEvent(BaseEvent):
+    """ Event representing the discovery of a PokeStop. """
```
Comment (kvangent) [inline PokeAlarm/Events/QuestEvent.py]: Update

File: PokeAlarm/Events/QuestEvent.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        # Used to reject
+        self.name = self.stop_id
+        self.geofence = Unknown.REGULAR
+        self.custom_dts = {}
+
+        # Quest Details
+        self.quest = data['quest']
+        self.reward = data['reward']
+        self.expiry = datetime.datetime.now().strftime("%d/%m/%Y 23:59")
```
Comment (kvangent) [inline PokeAlarm/Events/QuestEvent.py]: Does this field not get sent in the webhook? Might cause issues if processing is delayed (or sent around midnight or something) 

Also `expire` or `expire_time` would probably be better names for the field.

File: PokeAlarm/Events/QuestEvent.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        self.type = check_for_none(int, data.get('type'), 0)
```
Comment (kvangent) [inline PokeAlarm/Events/QuestEvent.py]: Might be some confusion with mon type - maybe `reward_type`?

File: PokeAlarm/Manager.py (hunk 2 above)
Comment (kvangent) [inline PokeAlarm/Manager.py]: This is still TODO?

File: PokeAlarm/Utilities/QuestUtils.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+                    j = j['quest_types']
+                    for id_ in j:
+                        nm = j[id_].lower()
+                        get_type.ids[nm] = int(id_)
+        if name in get_type.ids:
+            return get_type.ids[name]
+        else:
+            return int(name)  # try as an integer
+    except ValueError:
+        raise ValueError("Unable to interpret `{}` as a valid "
```
Comment (kvangent) [inline PokeAlarm/Utilities/QuestUtils.py:28]: This error message needs to be updated.

File: PokeAlarm/Utilities/QuestUtils.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+                         " severity name or id.".format(reward_type))
+    # if isinstance(type_id, int):
+    #     try:
+    #         if not hasattr(get_type, 'types'):
+    #             get_type.types = {}
+    #             files = glob(get_path('locales/*.json'))
+    #             for file_ in files:
+    #                 with open(file_, 'r') as f:
+    #                     j = json.loads(f.read())
+    #                     j = j['quest_types']
```
Comment (kvangent) [inline PokeAlarm/Utilities/QuestUtils.py]: Does this section still need to be here?

File: locales/zh_hk.json [hunk 14]
```diff
@@ -948,6 +948,16 @@
...
   },
+  "quest_types": {
+    "000": "Unset",
+    "001": "Experience",
+    "002": "Item",
+    "003": "Stardust",
+    "004": "Candy",
+    "005": "Avatar Clothing",
+    "006": "Quest",
+    "007": "Monster Encounter"
```
Comment (kvangent) [inline locales/zh_hk.json]: Probably should at least Google Translate these until people complain

File: prototype/events/quest.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    # Identification
+    stop_id = EventAttr(['pokestop_id'], str, required=True)  # type: str
+    type = EventAttr(['type'], int, default=0)  # type: int
+
+    # Details
+    quest = EventAttr(['quest'], str, required=True)  # type: str
+    reward = EventAttr(['reward'], str, required=True)  # type: str
+
+    @property
+    def expiry(self):
```
Comment (kvangent) [inline prototype/events/quest.py]: This property needs a better name.

File: prototype/events/quest.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+class Quest(Event):
+    """
+    Object representation of a change to a quest.
+    """
+
+    # Location
+    lat = EventAttr(['latitude'], float)  # type: float
+    lng = EventAttr(['longitude'], float)  # type: float
+    # Completed by Manager
+    distance = None  # type: float
```
Comment (kvangent) [inline prototype/events/quest.py:27]: Mark these with TODO please

File: start_pokealarm.py [hunk 17]
```diff
@@ -14,7 +14,7 @@
 import sys
 # 3rd Party Imports
 import configargparse
-from gevent import wsgi, spawn, signal, pool, queue
+from gevent import pywsgi, spawn, signal, pool, queue
```
Comment (kvangent) [inline start_pokealarm.py:17]: Why this change?

**Review by kvangent — CHANGES_REQUESTED** [review]
Also need wiki updates, but they can be in a separate PR if desired

File: PokeAlarm/Events/QuestEvent.py (hunk 10 above)
Comment (LegitDongo) [reply PokeAlarm/Events/QuestEvent.py]: It's not sent in the webhook
Quests always expire at midnight so there's not much point to send it.

File: start_pokealarm.py (hunk 17 above)
Comment (LegitDongo) [reply start_pokealarm.py:17]: gevent==1.2.2 is what we have in master, and 1.3.4 is in dev
As of gevent 1.3.0, wsgi was renamed to pywsgi
I probably should have just put this in a seperate PR 😅

File: PokeAlarm/Events/QuestEvent.py (hunk 10 above)
Comment (kvangent) [reply PokeAlarm/Events/QuestEvent.py]: Without sending the date, there is the possibility of introducing a bug where a webhook gets sent before midnight, is delayed for some reason, and isn't received until after. When PA displays the info it would then be incorrect. It's probably not a huge priority, but something to consider.

File: prototype/events/quest.py (hunk 15 above)
Comment (LegitDongo) [reply prototype/events/quest.py]: How about `expire_time`?

Comment (kvangent) [reply prototype/events/quest.py]: Sounds good to me.

File: PokeAlarm/Events/QuestEvent.py (hunk 10 above)
Comment (fosJoddie) [reply PokeAlarm/Events/QuestEvent.py]: Research Tasks change either at midnight or when Niantic starts/ends events that also changes the research task pool. 

For community days, this means research tasks will change 3 times during the day: midnight, 11 local time and finally back to the originals (src needed) at 14 local time.

For days that end or start events (like Halloween event) this means tasks rotate some arbitary time of the day, which you cannot easily know, since event times are special.

I don't see the point of implementing an expire_time at all for any purpose. The only reason why you'd want it is if PA queue is large and you somehow want PA to ignore too old data. This could be better done with a `datetime.datetime.now()` and check if it's still the same day.

File: PokeAlarm/Manager.py (hunk 2 above)
Comment (fosJoddie) [reply PokeAlarm/Manager.py]: The check has to check if both reward AND task (quest) is the same. Since two different tasks might give same reward.

File: PokeAlarm/Cache/Cache.py [hunk 18]
```diff
@@ -97,6 +98,12 @@ def day_or_night_id(self, s2_cell_id, day_or_night_id=Unknown.REGULAR):
             self._day_or_night_id[s2_cell_id] = day_or_night_id
         return self._day_or_night_id.get(s2_cell_id, Unknown.REGULAR)
 
+    def quest_reward(self, stop_id, reward=None):
```
Comment (fosJoddie) [inline PokeAlarm/Cache/Cache.py]: the uniqueness of quests is not only the stop_id + reward. There's also the quest/task itself.

the Cleanup method also needs some way to flush out all old rewards at midnight, or else pokestops that happen to have the same task/reward two days in a row will never get reported.

**Review by fosJoddie — CHANGES_REQUESTED** [review]

File: PokeAlarm/Events/QuestEvent.py (hunk 10 above)
Comment (kvangent) [reply PokeAlarm/Events/QuestEvent.py]: I still think this field should be sourced from the scanner side. I disagree that it doesn't have purpose - the field seems to have value for deduping, resolving the issues with delays, and with areas around horizons of time change. Presumably it's somewhat relevant for the user to know how long they have to pick up the quest as well?

Comment (LegitDongo) [reply PokeAlarm/Events/QuestEvent.py]: I can see how this is an issue on both sides, but using `datetime.datetime.now()` would force an expire time when it's possible the owner of the manual submissions may be aware of the varying expire times for quests on days such as community days. Likewise, there probably won't be much of a queue at to really have a need for an expire time field.

From where I'm sitting there's no easy way to solve both issues without adding another webhook field -- `override` or something like that -- which will tell PA to just ignore the cache.

Comment (fosJoddie) [reply PokeAlarm/Events/QuestEvent.py]: The protos in the real game suggests that they use an "updated" timestamp to figure out if a research task is the same or not. 
If the field is a "was discovered at TT" or "expected to expire at TT" doesn't make much difference to me, I'm just waffling out my own thoughts here.  Either way, if this field stays it will be more work for any tools wanting to send data to PA

File: PokeAlarm/Cache/FileCache.py [hunk 19]
```diff
@@ -42,6 +42,7 @@ def _load(self):
                 self._cell_weather_id = data.get('cell_weather_id', {})
                 self._severity_id = data.get('severity_id', {})
                 self._day_or_night_id = data.get('day_or_night_id', {})
+                self._reward = data.get('reward', {})
```
Comment (kvangent) [inline PokeAlarm/Cache/FileCache.py:45]: Looks like `_task` is missing?

File: PokeAlarm/Cache/FileCache.py [hunk 20]
```diff
@@ -64,7 +65,8 @@ def _save(self):
             'gym_image': self._gym_image,
             'cell_weather_id': self._cell_weather_id,
             'severity_id': self._severity_id,
-            'day_or_night_id': self._day_or_night_id
+            'day_or_night_id': self._day_or_night_id,
+            'reward': self._reward
```
Comment (kvangent) [inline PokeAlarm/Cache/FileCache.py]: `_task` here too?

File: PokeAlarm/Events/QuestEvent.py [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+        # Used to reject
+        self.name = self.stop_id
+        self.geofence = Unknown.REGULAR
+        self.custom_dts = {}
+
+        # Quest Details
+        self.quest = data['quest']
+        self.reward = data['reward']
+        self.expire_time = datetime.datetime.now().strftime("%d/%m/%Y 23:59")
```
Comment (kvangent) [inline PokeAlarm/Events/QuestEvent.py]: Where did we land on this? Seems like taking an argument with a default would be best?

**Review by kvangent — CHANGES_REQUESTED** [review]

**Review by kvangent — APPROVED** [review]
LGTM
